### ISSUE TITLE
403 Response on Namespace API

### ISSUE THREAD

**Issue by nagytech** [cloudstateio/cloudstate#123]
I just noticed, the cloudstate-operator is configured to watch all namespaces (https://github.com/cloudstateio/cloudstate/blob/14242147b6ab640517a71819514778bd1c3c57d3/operator/cloudstate.yaml#L65), but the ClusterRoleBinding linked to the role containing the `namespaces: list` permission is scoped to the `cloudstate` namespace.  

This results in a 403 when attempting to get the initial list of namespaces from the k8s api.

```
18:54:59.434 INFO  skuber.api - [ { reqId=b527729c-a34a-4e1c-9efd-ba13c0de89ba} } - about to send HTTP request: GET https://10.0.0.1/api/v1/namespaces]
18:54:59.551 INFO  skuber.api - [ { reqId=e5ec72b3-fc47-4c58-a517-d439783c3849} } - received response with HTTP status 200]
18:54:59.666 INFO  skuber.api - [ { reqId=b527729c-a34a-4e1c-9efd-ba13c0de89ba} } - received response with HTTP status 403]
18:54:59.728 INFO  skuber.api - [Response: non-ok status returned - Status(v1,Status,ListMeta(,,None),Some(Failure),Some(namespaces is forbidden: User "system:serviceaccount:default:cloudstate-operator" cannot list resource "namespaces" in API group "" at the cluster scope),Some(Forbidden),Some({"kind":"namespaces"}),Some(403))
18:54:59.748 WARN  a.s.s.RestartWithBackoffSource - Restarting graph due to failure. stack_trace: 
skuber.api.client.package$K8SException: Status(v1,Status,ListMeta(,,None),Some(Failure),Some(namespaces is forbidden: User "system:serviceaccount:default:cloudstate-operator" cannot list resource "namespaces" in API group "" at the cluster scope),Some(Forbidden),Some({"kind":"namespaces"}),Some(403))
	at skuber.api.client.impl.KubernetesClientImpl.$anonfun$toKubernetesResponse$1(KubernetesClientImpl.scala:648)
```

Since the default behaviour of k8s scoping is 'allow, deny', I can solve the problem by removing the namespace scope.  But, this would apply all other cloudstate-operator-role roles across each namespace - and the heightened permissions on apps, services and pods might not be required other namespaces.  So, I'm wondering if it's possible to delineate the roles as follows:

- roles over k8s resources required in all namespaces (eg. namespaces -> [get, list, watch])
- roles over cloudstate.io resources in all namespaces
- roles over all resources in the cloudstate namespace

That may look something like this:

```
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  name: namespace-listener
rules:

- apiGroups: [""]
  resources: ["namespaces"]
  verbs: ["get", "list", "watch"]

---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  name: cloudstate-io-manager
rules:

- apiGroups: ["cloudstate.io"]
  resources: ["statefulstores", "statefulservices"]
  verbs: ["get", "list", "watch"]

- apiGroups: ["cloudstate.io"]
  resources: ["statefulstores/status", "statefulservices/status"]
  verbs: ["update", "patch"]

---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  name: cloudstate-operator-role
rules:

- apiGroups: [""]
  resources: ["services", "pods"]
  verbs: ["get", "create", "delete", "patch", "update", "watch", "list"]

- apiGroups: ["apps"]
  resources: ["deployments", "deployments/scale"]
  verbs: ["get", "create", "delete", "patch", "update", "watch", "list"]

- apiGroups: ["rbac.authorization.k8s.io"]
  resources: ["rolebindings", "roles"]
  verbs: ["get", "create", "delete", "patch", "update"]

```

```
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: cloudstate-operator-namespace-listener
subjects:
- kind: ServiceAccount
  name: cloudstate-operator
roleRef:
  kind: ClusterRole
  name: namespace-listener
  apiGroup: rbac.authorization.k8s.io

---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: cloudstate-operator-cloudstate-io-manager
subjects:
- kind: ServiceAccount
  name: cloudstate-operator
roleRef:
  kind: ClusterRole
  name: cloudstate-io-manager
  apiGroup: rbac.authorization.k8s.io

---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: cloudstate-operator-role
subjects:
- kind: ServiceAccount
  name: cloudstate-operator
  namespace: cloudstate
roleRef:
  kind: ClusterRole
  name:  cloudstate-operator-role
  apiGroup: rbac.authorization.k8s.io

```

**Comment by viktorklang** [comment]
Thanks for the report, @nagytech!

Paging @jroper

**Comment by jroper** [comment]
Hey @nagytech, other people have reported finding this issue in private too, but I haven't been able to reproduce. It's a `ClusterRole` and `ClusterRoleBinding`, these are not namespaced resources, and in fact the `namespace` resource itself is also not namespaced (namespaces don't live in namespaces). The `namespace: cloudstate` in the `ClusterRoleBinding` was part of the `subject`, which in this case is the `cloudstate-operator` service account. That service account is in the `cloudstate` namespace, which is why `namespace: cloudstate` is needed there, otherwise the service account won't be able to be located. Unless I've misunderstood something.

I've actually just (finally) added smoke tests to CI for all this stuff in #124, that shows that it works in minikube on Kubernetes 1.15, and I've also tested it on GKE with Kubernetes 1.13. What Kubernetes distro and version are you using?

**Comment by jroper** [comment]
Actually, I think I understand now. The operator deployment is running in your default namespace, is that right?

So here's the problem, the CloudState YAML descriptor, for the most part, doesn't ever specify a namespace. This is because it doesn't want to require a particular namespace, though convention would be to create a namespaced called `cloudstate`, and then install it in that namepace by running `kubectl apply -n cloudstate -f cloudstate.yaml`. However, there is one place where it does configure the namespace, as you point out, in the `ClusterRoleBinding`, to refer to the `cloudstate-operator` service account that was created. But if that service account wasn't created in the `cloudstate` namespace, then that won't be valid, and so the permissions won't apply. I'm not sure what happens if you don't configure a namespace there, since `ClusterRoleBinding`'s are not namespaced, if you don't explicitly set a namespace, what does it default to? I'll need to experiment to find out, perhaps it defaults to the namespace that you deploy the `ClusterRoleBinding` to, even though its not namespaced? I don't know if that makes sense.

So to resolve your issue, either uninstall CloudState (`kubectl delete -f cloudstate.yaml`), then reinstall it in the `cloudstate` namespace, or, update the `ClusterRoleBinding` to list the namespace that you created the service account in (eg, `default`).

**Comment by jsravn** [comment]
One approach is to rely on kustomize (kubectl apply -k) - where you can specify a namespace in the kustomization.yaml file. This will correctly set the subject in the `ClusteRoleBinding` at install time. Benefit is you only need to set the namespace in one location.

**Comment by nagytech** [comment]
> Actually, I think I understand now. The operator deployment is running in your default namespace, is that right?

That's right.  Either I ran a deployment step out of order, or missed an instruction somewhere.

I'll run from scratch again to see if I can reproduce it - just trying to close off a few other things so might take me some time to get back to it.  Will just close this for now.

Also: running v1.14.6-gke.2 if that makes a difference.

**Comment by jroper** [comment]
@nagytech I just noticed that the installation instructions in the README don't mention the namespace, so that explains it. I've addressed that in #126.

These instructions are probably better:

https://cloudstate.io/docs/user/deployment/installing.html

However, they refer to version 0.4.3, which is the latest release, but there's another permissions issue in that. I'll hopefully push out a new release today that addresses that. Also, 0.4.3 is pre the name change from journal to statefulstore.
